Rakefile

```diff
@@ -1,16 +1,198 @@
-require 'bundler/gem_tasks'
-require 'rake/testtask'
-require 'rdoc/task'
+require 'appmap/version'
+GEM_VERSION = AppMap::VERSION
 
-begin
+namespace 'gem' do
+  require 'bundler/gem_tasks'
+end
+
+RUBY_VERSIONS=%w[2.5 2.6]
+FIXTURE_APPS=%w[rack_users_app rails_users_app]
+
+def build_base_image(ruby_version)
+  system "docker build"\
+         " --build-arg RUBY_VERSION=#{ruby_version} --build-arg GEM_VERSION=#{GEM_VERSION}"\
+         " -t appmap:#{GEM_VERSION} -f Dockerfile.appmap ." \
+  or raise 'Docker build failed'
+end
+  
+def build_app_image(app, ruby_version)
+  Dir.chdir "spec/fixtures/#{app}" do
+    system "env RUBY_VERSION=#{ruby_version} docker-compose build --build-arg RUBY_VERSION=#{ruby_version} --build-arg GEM_VERSION=#{GEM_VERSION}"\
+    or raise 'docker-compose build failed'
+  end
+end
+
+namespace :build do
+  namespace :base do
+    RUBY_VERSIONS.each do |ruby_version|
+      desc ruby_version
+      task ruby_version => ["gem:build"] do
+        build_base_image(ruby_version)
+      end.tap do |t|
+        desc "Build all images"
+        #multitask all: t
+        task all: t
+      end
+    end
+  end
+  
+  namespace :fixtures do
+    RUBY_VERSIONS.each do |ruby_version|
+      namespace ruby_version do
+        desc "build:fixtures:#{ruby_version}"
+        FIXTURE_APPS.each do |app|
+          desc app
+          task app => ["base:#{ruby_version}"] do
+            build_app_image(app, ruby_version)
+          end.tap do |t|
```

Comment:
Can you explain how this works, to define a task and then define another task (in this case `all), which may be defined multiple times?

Reply:
See my comment above, but let me know if that doesn't clear things up.